Migrate from homeconnect dependency to aiohomeconnect

Diegorro98 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Some programs at program switches (which are being deprecated) and program select entities may not be exposed anymore by the integration as they are missing at `aiohomeconnect` program key enumeration ([file](https://github.com/MartinHjelmare/aiohomeconnect/blob/ccfbf802884aa781eef481da92e6bccd2e6de733/src/aiohomeconnect/model/program.py#L213C1-L507C6)). You can get the diagnostics from configuration entry or the device and open an issue or submit a PR with the generated diagnostics file on the [`aiohomeconnect` repository](https://github.com/MartinHjelmare/aiohomeconnect) to suggest the missing program(s).

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Change the dependency on which Home Connect integrated is based on.
To achieve this change, many modifications were required, such using a data update coordinator, tests needed to adapt to how the new dependency works, new exceptions, data is stored in a different way, etc. So I achieved 100% of test coverage to make sure everything works fine.

The functionality remains the same. But the problem where the event stream closed abruptly and needed an integration reload has been finally fixed.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

This PR potentially fixes the following issues:

- #122506
- #122734
- #127938
- #130681
- #132659
- #135192

aiohomeconnect depends on httpx v0.28.1, and this repo didn't updated yet to that version at #133840 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Diegorro98 committed:
Migrate from homeconnect dependency to aiohomeconnect

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,227 @@
+                await asyncio.sleep(EVENT_STREAM_RECONNECT_DELAY)
+            except HomeConnectApiError as error:
+                _LOGGER.error("Error while listening for events: %s", error)
+                break
```
Here we probably want to reload the integration.

Diegorro98 replied:
How can we reload the integration from there?

MartinHjelmare replied:
If we have access to the config entry we can call `hass.config_entries.async_reload`. Example:

https://github.com/home-assistant/core/blob/b8632063f5d86931393abdc9840a41831a211470/homeassistant/components/matter/__init__.py#L187-L189

Diegorro98 replied:
Should I create the task also? or can I call `hass.config_entries.async_reload` directly?

MartinHjelmare replied:
Use this helper:

https://github.com/home-assistant/core/blob/75bdcee3e4921410485e3a99fbc8596036dfc5c6/homeassistant/config_entries.py#L2069

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,227 @@
+
+    async def _call_event_listener(self, event_message: EventMessage):
+        """Call listener for event."""
+        listeners = self._home_appliances_event_listeners.get(event_message.ha_id, {})
```
I'd look into using the coordinator context to selectively update the correct entities. The entities will already be connected to the coordinator if they inherit `CoordinatorEntity`.

We should update the coordinator state here in the coordinator from the event and then let the correct entities update their state.

MartinHjelmare replied:
Here's a example how to use the context:

https://github.com/home-assistant/core/blob/b8632063f5d86931393abdc9840a41831a211470/homeassistant/components/nibe_heatpump/coordinator.py#L25-L63

MartinHjelmare commented on homeassistant/components/home_connect/binary_sensor.py:
```diff
@@ -123,44 +122,49 @@ async def async_setup_entry(
+        self.set_native_value(cast(str | bool, event.value))
+        self.async_write_ha_state()
 
     async def async_update(self) -> None:
```
We shouldn't overwrite `async_update` in coordinator entities. It is defined by default and used to force a coordinator refresh via `homeassistant.update_entity` service action.

MartinHjelmare commented on homeassistant/components/home_connect/entity.py:
```diff
@@ -19,37 +20,52 @@ class HomeConnectEntity(Entity):
-        _LOGGER.debug("Entity update triggered on %s", self)
-        self.async_schedule_update_ha_state(True)
+    @abstractmethod
+    async def _async_event_update_listener(self, event: Event) -> None:
```
Inherit from `CoordinatorEntity` instead. It will have a method that is called by the coordinator on update.

MartinHjelmare commented on tests/components/home_connect/test_coordinator.py:
```diff
@@ -0,0 +1,298 @@
+    config_entry = MockConfigEntry(
+        state=ConfigEntryState.SETUP_IN_PROGRESS,
+    )
+    coordinator = HomeConnectCoordinator(hass, config_entry, client)
```
I don't think we should test the coordinator directly. It's a detail of the integration. It's tested indirectly by testing the entity state updates and config entry state.

Diegorro98 replied:
Yes, it is true that the normal functionality does not need to be tested specifically as it is already testes indirectly, but there are event stream edge cases that cannot be tested indirectly and should be tested to ensure that it works as expected.

MartinHjelmare replied:
What edge cases cannot be tested indirectly?

Diegorro98 replied:
- ignore `EventKey.UKNOWN` (tested at [`test_even_listener_ignore_unknowns`](https://github.com/home-assistant/core/blob/52ae1bbadd67dc9774bc2cab396fd3241dc5db97/tests/components/home_connect/test_coordinator.py#L134-L179))
- Events with settings or status that didn't have previously added to the coordinator data:
   this case happened while testing the changes and it happened when the settings and the status of an appliance could not be retrieved because the appliance was turned of, but when it turned on, events were received with status and settings that weren't previusly stored, so it raised an exceptions. Now the exception is handled and a there is a test ([`tests_receive_setting_and_status_for_first_time_at_events`](https://github.com/home-assistant/core/blob/52ae1bbadd67dc9774bc2cab396fd3241dc5db97/tests/components/home_connect/test_coordinator.py#L182-L230)) that checks that exception handling. 
- what happens when the event stream catches a Home Connect error:
  I test this at [`test_event_listener_error`](https://github.com/home-assistant/core/blob/52ae1bbadd67dc9774bc2cab396fd3241dc5db97/tests/components/home_connect/test_coordinator.py#L233-L246), where I check that the background task was simply terminated, but with your suggestion about requesting a configuration entry reload, I can also check that the cofiguration entry state is `setup_error`.
- Event stream abruptly ended:
  We have to make sure that, if the event stream abruptly ends, we are able to request the event stream, handle the events and deliver it to the entities again, like nothing happened there. (tested at [`test_event_listener_resilience`](https://github.com/home-assistant/core/blob/52ae1bbadd67dc9774bc2cab396fd3241dc5db97/tests/components/home_connect/test_coordinator.py#L249-L298))

MartinHjelmare replied:
All the mentioned cases can be tested by setting up the config entry and asserting the entity state and/or config entry state before and after the event(s).

Diegorro98 replied:
Yeah, for sure! (except ignore unknowns, where there is no entity that uses `EventKey.UNKNOW`, so I need to manually add a listener for  `EventKey.UNKNOW` to check that it has not been awaited)

Diegorro98 replied:
Btw, by saying "test directly" you mean the initialization of a coordinator instance and "test indirectly" you mean setting up the config entry?
Because I was thinking that "testing indirectly" meant to test through another tests outside `test_coordinator.py`.

MartinHjelmare replied:
Yes, I only mean that we shouldn't use the coordinator instance directly. We should use the Home Assistant core APIs and patched library to perform the test.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,227 @@
+                        await self.client.get_settings(appliance.ha_id)
+                    ).settings
+                }
+            except HomeConnectError as error:
```
If all appliances failed to fetch any settings or status we probably want to raise `UpdateFailed`.

Diegorro98 replied:
There are cases where is possible to get the appliance, but not the settings and the status because the appliance itself is turned off. So I think that those cases should not be considered a failed update, is just that is not in the correct state.

In future PRs (I already have the code ready) we could fetch the settings and the status of an appliance whenever the CONNECTED and PAIRED events are received, so that those cases eventually will get the appliance data on Home Assistant.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,227 @@
+        try:
+            appliances = await self.client.get_home_appliances()
+        except HomeConnectError as error:
+            raise UpdateFailed(
```
If we get a unauthorized error from the API here we want to start a reauth flow by raising `ConfigEntryAuthFailed`. This can be added in a follow up PR.

Diegorro98 committed:
Merge branch 'dev'

Diegorro98 committed:
Reload the integration if there is an API error on event stream

Diegorro98 committed:
fix typos at coordinator tests

Diegorro98 committed:
Setup config entry at coordinator tests

Diegorro98 committed:
fix ruff

Diegorro98 committed:
Bump aiohomeconnect to version 0.11.4

Diegorro98 committed:
Fix set program options

Diegorro98 committed:
Use context based updates at coordinator

Diegorro98 committed:
Improved how `context_callbacks` cache is invalidated

Diegorro98 committed:
fix

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -73,6 +75,29 @@ def __init__(
+        self, update_callback: CALLBACK_TYPE, context: Any = None
+    ) -> Callable[[], None]:
+        """Listen for data updates."""
+        remove_listener = super().async_add_listener(update_callback, context)
```
I think we need to invalidate the cached property also when adding a new listener, not only when removing a listener. See the nibe heatpump example.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -46,6 +47,7 @@ class HomeConnectApplianceData:
     info: HomeAppliance
     settings: dict[SettingKey, GetSetting]
     status: dict[StatusKey, Status]
+    events: dict[EventKey, Event] = field(default_factory=dict)
```
I'd sort the attributes :abcd:.

Diegorro98 replied:
You'd set `events` over `info`?
I feel like `info` is the frontline of the class so that it should be first (but `events` should be moved above `settings` for sure)

MartinHjelmare replied:
Just sort it :abcd:. Easier to find things then.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+
+        return remove_listener_and_invalidate_context_listeners
+
+    async def start_event_listener(self) -> None:
```
This can be an async callback since we don't await inside. Decorate it with `@callback`.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+            name=config_entry.entry_id,
+        )
+        self.client: HomeConnectClient = client
+        self._home_appliances_event_listeners: dict[
+            str, dict[EventKey, list[Callable[[Event], Coroutine[Any, Any, None]]]]
+        ] = {}
```
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/test_coordinator.py:
```diff
@@ -0,0 +1,319 @@
+    await hass.async_block_till_done()
+
+    listener.assert_called_once()
+    assert coordinator._home_appliances_event_listeners == {}
```
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/test_coordinator.py:
```diff
@@ -0,0 +1,319 @@
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    listener = MagicMock()
+    coordinator = cast(HomeConnectConfigEntry, config_entry).runtime_data
```
We should still remove the direct use of the coordinator here. I'd instead test a specific entity and check that the entity state is updated as long as the entity is added to Home Assistant and isn't updated after the entity is removed. An entity can be removed via the entity registry API.

Diegorro98 replied:
Got it. But if I remove the entity, then how do I check that the entity doesn't receive any updates (so that it keeps the same state)

MartinHjelmare replied:
Instead of removing the entity I think we can change the entity_id of the entity in the entity registry. This will remove the entity and then add it again. If we wouldn't clean up stale listeners there would be two listeners attached to the entity after the change of entity_id. We can listen for `EVENT_STATE_REPORTED` and check that the state isn't updated again with the same state.

Check that the logic checks out by commenting out the clean up and the test should fail.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+    async def _call_event_listener(self, event_message: EventMessage):
+        """Call listener for event."""
+        for event in event_message.data.items:
+            if event.key is EventKey.UNKNOWN:
```
Do we really need this case? There would need to be a listener registered for this key before it would matter. I don't think we will register a listener for this key, and if we would register a listener for this key, we probably want to see those events.

MartinHjelmare commented on tests/components/home_connect/test_coordinator.py:
```diff
@@ -0,0 +1,319 @@
+    event_key = EventKey.BSH_COMMON_STATUS_DOOR_STATE
+    listener = AsyncMock()
+    coordinator = cast(HomeConnectConfigEntry, config_entry).runtime_data
+    coordinator.async_add_listener(listener, (appliance_ha_id, event_key))
```
Just check the corresponding entity state instead.

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+                )
+                break
+
+    async def _call_event_listener(self, event_message: EventMessage):
```
```suggestion
    @callback
    def _call_event_listener(self, event_message: EventMessage):
```

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+                                    statuses[status_key] = Status(
+                                        status_key, event.value
+                                    )
+                            await self._call_event_listener(event_message)
```
```suggestion
                            self._call_event_listener(event_message)
```

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+                                        )
+                                else:
+                                    events[event.key] = event
+                            await self._call_event_listener(event_message)
```
```suggestion
                            self._call_event_listener(event_message)
```

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,226 @@
+                            events = self.data[event_message.ha_id].events
+                            for event in event_message.data.items:
+                                events[event.key] = event
+                            await self._call_event_listener(event_message)
```
```suggestion
                            self._call_event_listener(event_message)
```

Diegorro98 committed:
fixes and improvements at coordinator

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Diegorro98 committed:
Remove stale Entity inheritance

Diegorro98 committed:
Small improvement for light subscriptions

MartinHjelmare commented:
I'm testing this PR with my washer now. I'll report back tomorrow.

Diegorro98 commented:
Awesome, ty!

Diegorro98 committed:
Remove non-needed function

It had its purpose before some refactoring before the firs commit, no is no needed as is only used at HomeConnectEntity constructor

Diegorro98 committed:
Static methods and variables at conftest

MartinHjelmare commented:
I'm still waiting for the token refresh in some hours.

Some things I've thought of so far:

1. We should request a refresh of the coordinator to get the current state before continuing the event stream after an interruption, since we may have missed changes during the time the event stream was down.
2. We should include the `entity_id` in the state debug log in the entities.
3. Should we debounce the reloading of the config entry upon breaking error, by storing the last time of a breaking error in the config entry? This is just hypothetical at this point, but we could spam the API if the initial requests are successful but the event stream immediately breaks, causing us to reload the config entry and the cycle continues without delay.

MartinHjelmare commented:
The token refresh completed as expected. :+1:

Diegorro98 commented:
> 1. We should request a refresh of the coordinator to get the current state before continuing the event stream after an interruption, since we may have missed changes during the time the event stream was down.
> 2. We should include the `entity_id` in the state debug log in the entities. Eg: `"Updated %s, new state: %s", self.entity_id, state`

I agree with you, I will do it later today these changes, and an additional one: Fetch the appliance programs once at update, instead of doing it for select and switch platforms separately.

> 3. Should we debounce the reloading of the config entry upon breaking error, by storing the last time of a breaking error in the config entry? This is just hypothetical at this point, but we could spam the API if the initial requests are successful but the event stream immediately breaks, causing us to reload the config entry and the cycle continues without delay.

We only consider `HomeConnectApiError` exceptions as the only breaking errors, if there is an API error, it will happen also when `_async_update_data` is called at `async_config_entry_first_refresh` and there it will raise an `UpdateException` which will raise a config entry error.

The only thing I would add to event listener is a specific exception handler for `TooManyRequestsError`, which would wait for some seconds and then it would attempt again to stablish the event stream.

Diegorro98 commented:
Also, about API rate limits: Currently, a config entry can perform up to 50 -1 (for the appliances) request per minute, which, based on the current (or after the programs getter change) requests (settings, status, programs), we can request all the information we need right now for 16 appliances.
But in the future I'm planning to do more requests per appliance (settings, status, programs, active program, selected program), which allows to correctly fetch up to 9 appliances.
Should we worry about that?
I mean, we can get a higher number by delaying for a minute the update, but that would raise a warning log saying that the integration took about a minute to start... (and the Home Assistant complete startup will take also that time) so I think that's not the best solution.
Maybe the missing updates can be scheduled for a minute later?

thecode commented:
Hi,
I have tested this PR with a Bosch dryer, I had the integration in the past but it was removed for more than 2 years, still had the developer account active.

I am not sure the issues I am reporting are directly related to the dependency migration, so feel free to ignore if not relevant:

- Device names shows as "Import from configuration.yaml"
![image](https://github.com/user-attachments/assets/d80eafad-a744-479e-aa07-dc51b0eb0503)
I didn't have any YAML setting for this device
Deleting the device and re-adding it still keep this name

- Active program missing translation (using default English which should show translations)
![image](https://github.com/user-attachments/assets/b1f2e668-58d0-4235-98e6-49871191b8fd)

- Two door states are shown, one without translation:
![image](https://github.com/user-attachments/assets/009d06f3-b4d3-45e9-8f89-e35ef615330d)

- I am not sure what should be the relevant state of the control Switches and program select when `Remote Start` is disabled (I will discuss this unrelated to this PR)

Thanks.

MartinHjelmare commented:
> "Import from configuration.yaml"

This name comes from the application credentials entry. It's the name that was set when the application credentials for the integration was created. Your credentials were probably imported from YAML once and those credentials are still being used when the config entry is created today. You can check your existing application credentials in the GUI in the "more-info" menu in the top right corner of the Devices & services page.

Diegorro98 commented:
Jeez 🫠
can we leave that idea until someone reports that edge case? (so we can move forward with this PR also)

MartinHjelmare commented:
Yes, I don't think we need to change this now.

Diegorro98 committed:
Refresh the data after an event stream interruption

Diegorro98 committed:
Cleaned debug logs

Diegorro98 committed:
Fetch programs at coordinator

Diegorro98 committed:
Merge branch 'dev' into home_connect/aiohomeconnect

MartinHjelmare commented on homeassistant/components/home_connect/select.py:
```diff
@@ -211,80 +48,69 @@ async def async_setup_entry(
+        """Set the program value."""
+        event = self.appliance.events.get(cast(EventKey, self.bsh_key))
+        self._attr_current_option = (
+            PROGRAMS_TRANSLATION_KEYS_MAP.get(cast(ProgramKey, event.value), None)
```
```suggestion
            PROGRAMS_TRANSLATION_KEYS_MAP.get(cast(ProgramKey, event.value))
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -182,42 +181,45 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+        self.program = program
 
-    async def async_added_to_hass(self) -> None:
+    async def async_added_to_hass(self) -> None:  # pylint: disable=hass-missing-super-call
```
Why do we disable pylint here instead of solving the warning?

Diegorro98 replied:
Because the coordinator_context was based on the program key, so I didn't want the entity to be subscribed to a program key, it needed to be subscribed to a event key, specifically to the active program event key.
But I found another way to fix that, pushing it.

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -295,48 +294,34 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+    def update_native_value(self) -> None:
+        """Update the switch's status based on if the program related to this entity is currently active."""
+        event = self.appliance.events.get(EventKey.BSH_COMMON_ROOT_ACTIVE_PROGRAM)
+        if event and event.value == self.program.key:
             self._attr_is_on = True
         else:
             self._attr_is_on = False
```
```suggestion
        self._attr_is_on = bool(event and event.value == self.program.key)
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -383,46 +366,53 @@ async def async_turn_off(self, **kwargs: Any) -> None:
             self.power_off_state = BSH_POWER_STANDBY
         else:
             self.power_off_state = None
+        return
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -295,48 +294,34 @@ async def async_turn_off(self, **kwargs: Any) -> None:
 class HomeConnectPowerSwitch(HomeConnectEntity, SwitchEntity):
     """Power switch class for Home Connect."""
 
     power_off_state: str | None
```
Side note: I suggest we initialize the attribute to a sentinel value instead. It's bad practice in Python to have uninitialized attributes after the instance has been created unless there are some specific circumstances that require that.

https://github.com/home-assistant/core/blob/1face8df565e4ad16281bcfcf8dc7a280a15c203/homeassistant/helpers/typing.py#L31-L37

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -383,46 +366,53 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+        )
+        if not data or not data.constraints or not data.constraints.allowed_values:
+            try:
+                data = await self.coordinator.client.get_setting(
```
I don't understand why we need to fetch this explicitly here. We seem to be asking for the data for the `SettingKey.BSH_COMMON_POWER_STATE` which is the same key used for the state of this entity. This entity isn't added unless that key exist in the coordinator state. We have that check above in `async_setup_entry`. Why wouldn't all data be available before creating the entity if the key is there?

MartinHjelmare replied:
I'd fetch the settings before creating and adding the entity so we don't need to do that later.

MartinHjelmare replied:
We don't need to change this in this PR though.

Diegorro98 replied:
I think it is better where it is already, because fetching it here allows us to fetch it only if the power setting is present in the appliance (this can be also done at wherever we would move it) and if the entity is enabled.

We could even move it to the turn off method so that we fetch it only when it is strictly necessary (although it might consume some "user time" instead of doing it while the user doesn't explicitly need it)

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -110,41 +106,48 @@ async def async_setup_entry(
     """Set up the Home Connect switch."""
 
-    def get_entities() -> list[SwitchEntity]:
+    def get_entities_for_appliance(
```
We don't need this function since it's only called once. We can inline the code.

Diegorro98 commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -0,0 +1,254 @@
+                translation_placeholders=get_dict_from_home_connect_error(error),
+            ) from error
+
+        appliances_data = self.data or {}
```
it is a `dict[str, HomeConnectApplianceData]`, can it be set to `defaultdict(dict)`?
Is the first time that I use `defaultdict`, so I don't know the capabilities, but AFAIK, the type that is inside the parenthesis (in this case `dict`) is the type of the dictionary values (which actually is `HomeConnectApplianceData`).
Shouldn't it be `defaultdict(HomeConnectApplianceData)`? And if it was, it wouldn't for default cases because it raises this exception: `TypeError: HomeConnectApplianceData.__init__() missing 3 required keyword-only arguments: 'info', 'settings', and 'status'`

Diegorro98 committed:
Improvements

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -383,46 +366,53 @@ async def async_turn_off(self, **kwargs: Any) -> None:
-        if not data or not (
-            allowed_values := data.get(ATTR_CONSTRAINTS, {}).get(ATTR_ALLOWED_VALUES)
-        ):
+        data = self.appliance.settings.get(
+            SettingKey.BSH_COMMON_POWER_STATE,
+        )
+        if not data or not data.constraints or not data.constraints.allowed_values:
```
Based on the discussion below this should be enough since we know that the key is in the settings but we don't know if the data has the constraints.
```suggestion
        data = self.appliance.settings[SettingKey.BSH_COMMON_POWER_STATE]
        
        if not data.constraints or not data.constraints.allowed_values:
```

Diegorro98 commented on homeassistant/components/home_connect/diagnostics.py:
```diff
@@ -2,50 +2,69 @@
+        # That's why here  we fetch the programs with a raw response so we can
+        # get the actual program keys and the user can suggest the addition
+        # of the missing programs to the enum to the aiohomeconnect library.
+        program_response = await client._auth.request(  # noqa: SLF001
```
How do we populate raw_key? I was thinking about overwritting `from_dict` method to populate it, but it is decorated as `@final`

MartinHjelmare replied:
Mashumaro should have some interface where you can do custom things during deserialization and validation. Check those docs/readme.

MartinHjelmare replied:
Maybe it's enough to use the alias `"key"` for this attribute?

Diegorro98 committed:
Simplify obtaining power settings from coordinator data

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Diegorro98 committed:
Remove unnecessary statement

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -383,46 +348,51 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+            and value == self.power_off_state
         ):
             self._attr_is_on = False
+        elif isinstance(self.power_off_state, UndefinedType) and value in [
```
```suggestion
        elif self.power_off_state is UNDEFINED and value in [
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -295,48 +279,31 @@ async def async_turn_off(self, **kwargs: Any) -> None:
+        """Call when entity is added to hass."""
         await super().async_added_to_hass()
-        if not hasattr(self, "power_off_state"):
+        if isinstance(self.power_off_state, UndefinedType):
```
```suggestion
        if self.power_off_state is UNDEFINED:
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -345,19 +312,18 @@ async def async_turn_on(self, **kwargs: Any) -> None:
     async def async_turn_off(self, **kwargs: Any) -> None:
         """Switch the device off."""
-        if not hasattr(self, "power_off_state"):
+        if isinstance(self.power_off_state, UndefinedType):
```
```suggestion
        if self.power_off_state is UNDEFINED:
```

Diegorro98 committed:
use `is UNDEFINED` instead of `isinstance`

Diegorro98 committed:
Request power setting only when it is strictly necessary

Diegorro98 committed:
Bump aiohomeconnect to 0.12.1

Diegorro98 committed:
use raw keys for diagnostics

MartinHjelmare commented on homeassistant/components/home_connect/coordinator.py:
```diff
@@ -143,7 +143,7 @@ async def _event_listener(self) -> None:
                                     else:
                                         settings[setting_key] = GetSetting(
-                                            setting_key, event.value
+                                            setting_key, setting_key.value, event.value
```
I think we should use keyword arguments for the models now so it's easier to read and understand the interface.

MartinHjelmare replied:
Side note: Maybe we should even require that in the models so it'd easier to add new optional parameters without a breaking change?

Diegorro98 committed:
Use keyword arguments where needed

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented on tests/components/home_connect/test_switch.py:
```diff
@@ -58,231 +56,287 @@ def platforms() -> list[str]:
+    assert await integration_setup(client)
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    # appliance.status.update(status)
```
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/test_switch.py:
```diff
@@ -58,231 +56,287 @@ def platforms() -> list[str]:
-    await hass.services.async_call(
-        SWITCH_DOMAIN, service, {ATTR_ENTITY_ID: entity_id}, blocking=True
-    )
+    # appliance.status.update(status)
```
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/test_coordinator.py:
```diff
@@ -0,0 +1,372 @@
+    await integration_setup(client_with_exception)
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    assert cast(HomeConnectConfigEntry, config_entry).runtime_data.data
```
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/test_coordinator.py:
```diff
@@ -0,0 +1,372 @@
+    await integration_setup(client)
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    assert cast(HomeConnectConfigEntry, config_entry).runtime_data.data
```
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/test_init.py:
```diff
@@ -156,72 +144,61 @@ async def test_api_setup(
+        client_mock.side_effect = MagicMock(side_effect=init_side_effect)
+        assert await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+    config_entry.runtime_data.client = client
```
This shouldn't be needed if we patch correctly. We shouldn't touch or use `runtime_data` in the tests.
```suggestion
```

MartinHjelmare commented on tests/components/home_connect/conftest.py:
```diff
@@ -102,158 +121,231 @@ def platforms() -> list[Platform]:
             result = await hass.config_entries.async_setup(config_entry.entry_id)
             await hass.async_block_till_done()
+        if result:
+            config_entry.runtime_data.client = client
```
This needs to go.

MartinHjelmare reviewed (changes requested):
Sorry, I found some things in the tests that need to change.

Diegorro98 committed:
Remove unnecessary statements

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Thanks!
